src/validationinterface.h

```diff
@@ -31,6 +32,11 @@ void RegisterValidationInterface(CValidationInterface* pwalletIn);
 void UnregisterValidationInterface(CValidationInterface* pwalletIn);
 /** Unregister all wallets from core */
 void UnregisterAllValidationInterfaces();
+/**
+ * Pushes a function to callback onto the notification queue, guaranteeing any
+ * callbacks generated prior to now are finished when the function is called.
+ */
+void CallFunctionInValidationInterfaceQueue(std::function<void ()> func);
```

Comment:
I think a better interface would be
`std::future<void> SyncWithValidationInterfaceQueue();`

Whose implementation sets up the delayed call to a `promise.set_value()`, instead of this being the responsibility of the caller as it is now.

The sync functionality is the only purpose, so rather than have an overly-generic interface which allows calling any function, it's better to have a more specific interface which simplifies the interface as well as the setup work on the caller side.

Reply:
Hmm..I have a few more branches which use this function for other stuff (eg one that will set a "allowed to continue processing" boolean on a CNodeState then call WakeMessageHandler). I think most of the use-cases for now could almost do fine with a future-like return value if std::future allowed you to query whether the future had yet completed. Still, since the promise-create-future-wait pattern isnt too many LoC, I suppose I'll leave it for now instead of adding a bunch of machinery just to create a special future-like return type.

Reply:
> I think most of the use-cases for now could almost do fine with a future-like return value if `std::future` allowed you to query whether the future had yet completed.
Wouldn't `std::future::wait_for()` with argument provide that functionality?

Reply:
Yea, could pull some hack like wait_for(1ns), but that's pretty gross. I think the purpose is somewhat of both - really it shouldn't be to do much more than sync, but sync may require rather arbitrary things (like calling WakeMessageHandler to wake up another thread which is waiting on some other unrelated condition in the example above). For now probably best to just leave it IMO.

Reply:
Ok.
btw I meant to write "with argument *zero*" above, so it's not hacky - it just checks if the `std::future` is done, without waiting.